extensions/amp-share-tracking/0.1/amp-share-tracking.js

```diff
@@ -15,41 +15,173 @@
+  /**
+    * @return {boolean}
+    * @private
+    */
+  isExperimentOn_() {
+    return isExperimentOn(this.getWin(), TAG);
+  }
+
+  /** @override */
+  isLayoutSupported(unusedLayout) {
+    return true;
+  }
+
+  /** @override */
+  createdCallback() {
+    if (!this.isExperimentOn_()) {
+      dev.warn(TAG, `TAG ${TAG} disabled`);
+      return;
+    }
+
+    /** @private @const {!Window} */
+    this.win_ = this.getWin();
+
+    /** @private @const {!ShareTrackingService} */
+    this.shareTrackingService_ = getShareTrackingService_(this.win_);
+  }
 
   /** @override */
   buildCallback() {
-    this.isExperimentOn_ = isExperimentOn(this.getWin(), TAG);
-    if (!this.isExperimentOn_) {
+    if (!this.isExperimentOn_()) {
       dev.warn(TAG, `TAG ${TAG} disabled`);
       return;
     }
+
+    this.vendorHref = this.element.getAttribute('data-href');
+    dev.fine(TAG, 'vendorHref: ', this.vendorHref);
+
+    return Promise.all([
+      this.shareTrackingService_.getIncomingFragment(),
+      this.shareTrackingService_.getOutgoingFragment(this.vendorHref)]).then(
+        results => {
+          this.incomingFragment = results[0];
+          this.outgoingFragment = results[1];
+          dev.fine(TAG, 'get incoming fragment: ' + this.incomingFragment);
+          dev.fine(TAG, 'get outgoing fragment: ' + this.outgoingFragment);
+        });
   }
 }
 
 /**
- * ShareTrackingService processes the incoming and outcoming url fragment
+ * ShareTrackingService processes the incoming and outgoing url fragment
  */
 export class ShareTrackingService {
   constructor(window) {
-    this.win = window;
+    this.win_ = window;
   }
+
+  /**
+   * Get the incoming share-tracking fragment from the url
+   * @return {!Promise<string>}
+   */
+  getIncomingFragment() {
+    // remove the #
+    const hash = this.win_.location.hash.substr(1);
+
+    // when the url contains the share-tracking identifier directly
+    if (hash.indexOf('.') == 0) {
+      const endIndex = hash.indexOf('&');
+      if (endIndex != -1) {
+        return Promise.resolve(hash.substr(1, (endIndex - 1)));
+      } else {
+        return Promise.resolve(hash.substr(1));
+      }
+    }
+
+    // get the share-tracking identifier from the viewer
+    return viewerFor(this.win_).getShareTrackingIncomingFragment();
+  }
+
+  /**
+   * Get an outgoing share-tracking fragment
+   * @return {!Promise<string>}
+   */
+  getOutgoingFragment(vendorUrl) {
+    if (vendorUrl) {
+      return this.getOutgoingFragmentFromVendor(vendorUrl).then(
+        response => {
+          if (!response.outgoingFragment) {
+            dev.warn(TAG, 'The response from [' + vendorUrl +
+                '] does not have an outgoingFragment value. ' +
+                'Generating outgoing fragment using random generator.');
+            return this.getOutgoingFragmentRandomly();
+          } else {
+            return Promise.resolve(response.outgoingFragment);
```

Comment:
If `getOutgoingFragment` returns a string instead of a Promise<string>, then the caller `buildCallback` cannot use `.then` to determine that retrieving fragment is done. Or can it?